solidity/contracts/liquidity-protection/TokenTimeWeightedAverage.sol

```diff
@@ -0,0 +1,375 @@
+        mapping(uint256 => Fraction) accumulators;
+        mapping(uint256 => bool) timestamps;
+    }
+
+    uint256 private constant INITIAL_SAMPLE_N = 0;
+    uint256 private constant INITIAL_SAMPLE_D = 1;
+    uint256 private constant MAX_UINT128 = 2**128 - 1;
+
+    mapping(IERC20Token => TokenData) private data;
+
+    // the owner role is used to add values to the accumulator, but it can't update them
+    bytes32 public constant ROLE_OWNER = keccak256("ROLE_OWNER");
+
+    // the seeder roles is used to seed the accumulator with past values
+    bytes32 public constant ROLE_SEEDER = keccak256("ROLE_SEEDER");
+
+    /**
+     * @dev triggered when an accumulator for a specific token is being initialized
+     *
+     * @param _token the token the data is accumulated for=
+     * @param _startTime the accumulation starting time
+     */
+    event Initialized(IERC20Token indexed _token, uint256 _startTime);
+
+    /**
+     * @dev triggered when a new sample is being added
+     *
+     * @param _token the token the data is accumulated for
+     * @param _n ratio numerator
+     * @param _d ratio denominator
+     * @param _time the sampling timestamp
+     */
+    event SampleAdded(IERC20Token indexed _token, uint256 _n, uint256 _d, uint256 _time);
+
+    constructor() public {
+        // set up administrative roles.
+        _setRoleAdmin(ROLE_OWNER, ROLE_OWNER);
+        _setRoleAdmin(ROLE_SEEDER, ROLE_OWNER);
+
+        // allow the deployer to initially govern the contract.
+        _setupRole(ROLE_OWNER, msg.sender);
+    }
+
+    modifier initialized(IERC20Token _token) {
+        _initialized(_token);
+        _;
+    }
+
+    // error message binary size optimization
+    function _initialized(IERC20Token _token) internal view {
+        require(isInitialized(_token), "ERR_NOT_INITIALIZED");
+    }
+
+    /**
+     * @dev initializes the accumulator for a specific token
+     * can only be called by an owner or a seeder
+     *
+     * @param _token the token the data is accumulated for
+     * @param _startTime the accumulation starting time
+     */
+    function initialize(IERC20Token _token, uint256 _startTime) external override validAddress(address(_token)) {
+        require(hasRole(ROLE_OWNER, msg.sender) || hasRole(ROLE_SEEDER, msg.sender), "ERR_ACCESS_DENIED");
+        require(_startTime <= time(), "ERR_INVALID_TIME");
+
+        // make sure that we're not attempting to initialize the accumulator twice
+        require(!isInitialized(_token), "ERR_ALREADY_INITIALIZED");
+
+        addSample(_token, Fraction({ n: INITIAL_SAMPLE_N, d: INITIAL_SAMPLE_D }), _startTime);
+
+        emit Initialized(_token, _startTime);
+    }
+
+    /**
+     * @dev adds a new sample to the accumulator
+     * can only be called by an owner
+     *
+     * @param _token the token the data is accumulated for
+     * @param _n ratio numerator
+     * @param _d ratio denominator
+     */
+    function addSample(
```

Comment:
Code reuse: you can simply call `addPastSample(_n, _d, time())` here.
I would rename `addPastSample` to `addSample` (i.e., function-overload with this one), because `time()` means that we're adding a present-sample, not a past-sample (I apologize if it sounds like an English grammar lesson).

Reply:
In both of these cases, I wanted to avoid triggering the input validating modifiers twice, so both reuse the same `private` implementation